src/main/java/frc/robot/subsystems/drive/Drivetrain.java

```diff
@@ -1,16 +1,339 @@
+   */
+  public void drive(Vector spdVec, Vector rotVec, boolean altDrive, boolean centerGyro) {
+    if (centerGyro) zeroHeading();
+    if (altDrive) {
+      altDrive(spdVec, rotVec);
+    } else {
+      mainDrive(spdVec, rotVec.x());
+    }
+  }
+
+  /**
+   * moves the drivetrain using the main turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the speed to rotate with (-1, 1)
+   */
+  private void mainDrive(Vector spdVec, double xRot) {
+    double rot = xRot * DriveConfig.kMaxAngularSpeed;
+    move(spdVec, rot);
+  }
+
+  /**
+   * gets the value of the robot's gyro as a Measure<Angle>
+   *
+   * @see Measure
+   * @return the angle of the robot gyro
+   */
+  private Measure<Angle> getGyroAngle() {
+    return Units.Degrees.of(m_gyro.getAngle());
+  }
+
+  /**
+   * moves the drivetrain using the alternative turning mode
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param xRot the x component of the direction vector to point towards
+   * @param yRot the y component of the direction vector to point towards
+   */
+  private void altDrive(Vector spdVec, Vector rotVec) {
+    double rot = 0;
+    m_rightAngGoalRadians = rotVec.angle();
+    if (rotVec.squaredMag() > 0) {
+      double stickAng = m_rightAngGoalRadians;
+      // gets the difference in angle, then uses mod to make sure its from -PI rad to PI rad
+      rot = altTurnSmooth(stickAng);
+    }
+    m_turnDirRadians = rot;
+    move(spdVec, rot);
+  }
+
+  private double altTurnSmooth(double stickAng) {
+    return Math.tanh(
+            ((getGyroAngle().in(Units.Radians) + stickAng + Math.PI) % (2 * Math.PI) - Math.PI)
+                / DriveConfig.altTurnSmoothing)
+        * DriveConfig.kMaxAngularSpeed;
+  }
+
+  /**
+   * moves the drivetrain using the given values
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param rot the angular velocity to rotate the drivetrain in radians/s
+   */
+  private void move(Vector spdVec, double rot) {
+    move(spdVec, rot, true);
+  }
+
+  /**
+   * moves the drivetrain using the given values
+   *
+   * @param xSpeed the proportion of the robot's max velocity to move in the x direction
+   * @param ySpeed the proportion of the robot's max velocity to move in the y direction
+   * @param rot the angular velocity to rotate the drivetrain in radians/s
+   * @param rateLimit whether or not to use slew rate limiting
+   */
+  private void move(Vector spdVec, double rot, boolean rateLimit) {
+    Vector spdCommanded = spdVec;
+
+    spdVec.rot(0);
```

Comment:
This is a hack -- let's find the real root cause of this problem instead of patching it like this.

Reply:
I think I found and fixed it in the most recent drivetrain commit (f6251eb), but we should test it this meeting if possible

Reply:
we found that the root of the problem was in the slew rate limiting code, and I think it was an issue of Vector rotating the wrong way, but we'll have to test again to be sure